### PR TITLE
Lf 3955 the white screen is displayed once the user deletes a crop plan for a wild crop with the transplant task in it and clicks on the button 2 times

### PR DESCRIPTION
**Pull request by ebertmota**
**Description**

**About the bug:**
For some reason when you create a crop plan answering these questions like this:
```
“Will you be planting this crop or is it already in the ground?” - In ground
“Are you harvesting a wild crop?” - Yes
“Will this crop be transplanted?” - Yes
```
After the creation is complete the first route of the creation flow: “add_management_plan/planted_already” stays on the browser history, and, when the user tries to go back using the browser go back button or the LiteFarm internal go back button, instead of being redirected to the crop management route the user is redirected to the planted_already route.

**About the solution:**
I was not able to identify what is causing this bug, so the changes that I've made are not the best way to fix this issue but minimize it by redirecting the user to the correct route and preventing the user from viewing an empty white screen.

Jira link: https://lite-farm.atlassian.net/browse/LF-3955

**Type of change**

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

Steps to reproduce:
1. Open the website  
2. Login with your account
3. Click on the Crops > Select any crop 
4. Click on the “Add a plan” button
5. Will you be planting this crop or is it already in the ground?  In Ground
6. Put Any Age 
7. Are you harvesting a wild crop? Yes 
8. Click on the “Continue” button
9. Will this crop be transplanted? Yes 
10. Complete the creation flow.
11. Click on the browser go-back button or the internal LiteFarm go-back button component “<“based on the Header.

- [ ] Passes test case
- [x] UI components visually reviewed on desktop view
- [ ] UI components visually reviewed on mobile view
- [ ] Other (please explain)

**Checklist:**

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have added "MISSING" for all new language tags to languages I don't speak
- [ ] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

### CODE REVIEW

**Review by SayakaOno — COMMENTED** [review]
Thank you for tackling this, your solution looks great as a workaround!!

> “add_management_plan/planted_already” stays on the browser history

I tried investigating this, and as you mentioned, I found that history manipulation was not working correctly.

Let me try to explain what the actual issue was...

We have the hook `useHookFormPersist` that takes care of page navigation/manipulation for a multi-page form. The logic is that whenever the user gets out of the form (= if the new path is not included in `persistedPaths`), we remove the pages in the form by doing `history.go(-x)` where `x` is calculated based on `historyStack` in Redux store.

The problem in the ticket is as you found, the first page of the form is not removed from browser's history, which means historyStack's calculation is wrong.

So, I found that one path is missing in `historyStack`, and it is `/crop/<id>/add_management_plan/choose_final_planting_location`.
Screenshot 2024-01-19 at 10 55 20 AM
It happens because the same component is used in the previous page `.../choose_initial_planting_location`, and re-render doesn't happen in `.../choose_final_planting_location`, so `useHookFormPersist` is not called in `packages/webapp/src/components/Crop/PlantingLocation/index.jsx` (line 35), and the path is not added to `historyStack`.

The solution is to make re-render happen in `PlantingLocation`, and you can do so by passing `key` to the component. When `key` changes, the component is re-rendered.


I hope it makes sense... please let me know if you have any questions!

**Comment by ebertmota** [comment]
> Thank you for tackling this, your solution looks great as a workaround!!
> 
> > “add_management_plan/planted_already” stays on the browser history
> 
> I tried investigating this, and as you mentioned, I found that history manipulation was not working correctly.
> 
> Let me try to explain what the actual issue was...
> 
> We have the hook `useHookFormPersist` that takes care of page navigation/manipulation for a multi-page form. The logic is that whenever the user gets out of the form (= if the new path is not included in `persistedPaths`), we remove the pages in the form by doing `history.go(-x)` where `x` is calculated based on `historyStack` in Redux store.
> 
> The problem in the ticket is as you found, the first page of the form is not removed from browser's history, which means historyStack's calculation is wrong.
> 
> So, I found that one path is missing in `historyStack`, and it is `/crop/<id>/add_management_plan/choose_final_planting_location`. Screenshot 2024-01-19 at 10 55 20 AM It happens because the same component is used in the previous page `.../choose_initial_planting_location`, and re-render doesn't happen in `.../choose_final_planting_location`, so `useHookFormPersist` is not called in `packages/webapp/src/components/Crop/PlantingLocation/index.jsx` (line 35), and the path is not added to `historyStack`.
> 
> The solution is to make re-render happen in `PlantingLocation`, and you can do so by passing `key` to the component. When `key` changes, the component is re-rendered.
> 
> I hope it makes sense... please let me know if you have any questions!

Thanks 

> Thank you for tackling this, your solution looks great as a workaround!!
> 
> > “add_management_plan/planted_already” stays on the browser history
> 
> I tried investigating this, and as you mentioned, I found that history manipulation was not working correctly.
> 
> Let me try to explain what the actual issue was...
> 
> We have the hook `useHookFormPersist` that takes care of page navigation/manipulation for a multi-page form. The logic is that whenever the user gets out of the form (= if the new path is not included in `persistedPaths`), we remove the pages in the form by doing `history.go(-x)` where `x` is calculated based on `historyStack` in Redux store.
> 
> The problem in the ticket is as you found, the first page of the form is not removed from browser's history, which means historyStack's calculation is wrong.
> 
> So, I found that one path is missing in `historyStack`, and it is `/crop/<id>/add_management_plan/choose_final_planting_location`. Screenshot 2024-01-19 at 10 55 20 AM It happens because the same component is used in the previous page `.../choose_initial_planting_location`, and re-render doesn't happen in `.../choose_final_planting_location`, so `useHookFormPersist` is not called in `packages/webapp/src/components/Crop/PlantingLocation/index.jsx` (line 35), and the path is not added to `historyStack`.
> 
> The solution is to make re-render happen in `PlantingLocation`, and you can do so by passing `key` to the component. When `key` changes, the component is re-rendered.
> 
> I hope it makes sense... please let me know if you have any questions!

Thank you so much for the explanation!

I understand the issue now, and I've made the change that you suggested, adding a key to the PlantingLocation component and it fixed the problem

File: packages/webapp/src/components/Crop/ManagementDetail/ManagementPlanTasks.jsx [hunk 1]
```diff
@@ -75,7 +75,7 @@ export default function PureManagementTasks({
         )
       }
     >
-      <CropHeader onBackClick={() => history.go(-1)} variety={variety} />
+      <CropHeader onBackClick={onBack} variety={variety} />
```
[the remark this answers is not in the document]
Comment (ebertmota) [reply packages/webapp/src/components/Crop/ManagementDetail/ManagementPlanTasks.jsx]: Make sense, I've undone this change

**Review by SayakaOno — APPROVED** [review]
Thank you for updating, and I'm sorry for the late re-review.
The fix looks good to me!

**Comment by antsgar** [comment]
@ebertmota thank you so much for working on this! This bug definitely supports the notion that we need to refactor the multistep form/navigation logic as soon as possible. We shouldn't be forcing re-renders of a component and instead let React do its thing, so having to do this is a sign that something's inherently wrong with how we're handling the navigation (it's something that we're currently researching and looking to refactor). That being said I'm alright with this fix for now since it's the best we'll have at the moment, but could we add a comment above the change with a TODO indicating we should revert it back once navigation is refactored please? 🙏

**Comment by kathyavini** [comment]
@SayakaOno it was `src/containers/Finances/ExpenseDetail/`, and the pattern was copied for `src/containers/Finances/RevenueDetail/`. And checked it just now and you can indeed get into a deletion + back button crash (read > edit > (sends you back to transaction list) > read > delete > back > back), is that the same cause??!

**Comment by SayakaOno** [comment]
Thank you @kathyavini! I wasn't expecting a crash there 😅 
I can't always reproduce it, and I don't think it's related to this bug, but it should be related to the hook...
A few things seem strange to me, and this is one of them:
Screenshot 2024-01-26 at 3 45 14 PM
Going to the read-only view adds its path to `historyStack` and then going to the edit view adds its path to `persistedPaths`.

**Comment by ebertmota** [comment]
> @ebertmota thank you so much for working on this! This bug definitely supports the notion that we need to refactor the multistep form/navigation logic as soon as possible. We shouldn't be forcing re-renders of a component and instead let React do its thing, so having to do this is a sign that something's inherently wrong with how we're handling the navigation (it's something that we're currently researching and looking to refactor). That being said I'm alright with this fix for now since it's the best we'll have at the moment, but could we add a comment above the change with a TODO indicating we should revert it back once navigation is refactored please? 🙏

Thanks @antsgar!
I've included the TODO comment that you suggested indicating that we need to revert this change after the navigation refactor

**Review by SayakaOno — APPROVED** [review]

**Review by antsgar — APPROVED** [review]

**Pull request merged**